### ISSUE TITLE
Doesn't support variables assigned to "this" in the beforeEach for acceptance tests

### ISSUE THREAD

**Issue by mdentremont** [elwayman02/ember-sinon-qunit#25]
With ember-qunit, in my `beforeEach` I can assign to this, and these assigned attributes will be available inside my test callbacks, ex:

```
moduleForComponent(..., function () {
  beforeEach() {
    this.demo = 'string';
});

test('test name', function (assert) {
  assert.equal(this.demo, 'string');
});
```

The following works with ember-qunit, but not when using ember-sinon-qunit.

Ideally ember-sinon-qunit would work the same as ember-qunit, so replace all usages of ember-qunit with the sinon variant.

**Comment by elwayman02** [comment]
Yea, I just noticed this as well. I'm not sure what's going on exactly, but the context is being lost somewhere. Thanks @mdentremont!

**Comment by elwayman02** [comment]
It would really help if you could submit a PR adding a failing test for this scenario, so I can track the problem and get a fix out!

**Comment by elwayman02** [comment]
@rwjblue: any idea if this issue is caused by something upstream in ember-cli-qunit or ember-test-helpers? I haven't had a chance to dig in yet.

**Comment by BrianSipple** [comment]
It looks like the reason this is happening is that the context of `test` from `ember-sinon-qunit` is the window, whereas the context in the `beforeEach` and `afterEach` hooks of `moduleForComponent`, `moduleForAcceptance`, etc -- alongside the `test` function exported by `ember-qunit` -- have all been [wired up to be the current application container](https://github.com/switchfly/ember-test-helpers/blob/master/lib%2Fember-test-helpers%2Ftest-module.js#L107). 

Hence the mismatch.

[I PR'd a workaround](https://github.com/elwayman02/ember-sinon-qunit/pull/50) for this that would allow a user of `ember-sinon-qunit` to set properties on a `context` object belonging to the imported `test` function instance, and then access it on `this` inside of the `test` function as they normally would inside of `ember-qunit`'s `test` function. 

Part of me feels that this approach is a _bit_ inelegant, so I'm certainly not partial to it. That said, it did seem a lot more straightforward than trying to interface with the context set up by `ember-qunit`'s module helpers, and I figured I'd get the PR out there for discussion as I was going through and figuring out the root cause of the issue.

**Comment by BrianSipple** [comment]
Interestingly, I can't seem to replicate this issue any more after updating to 1.3.3. @mdentremont & @elwayman02, can you confirm that this behavior no longer exists there. I can close my PR if so.

**Comment by elwayman02** [comment]
@BrianSipple I have a skipped test that shows this failing for regular `qunit` modules, but I have long suspected that the primary issue was upstream in `ember-cli-qunit` itself. The context problem is definitely a separate issue in each case, as `qunit` probably handles context-passing differently, and `ember-cli-qunit` likely purposefully preserves the context from `beforeEach` to `test`. I wonder if we can add a test showing this working for a `moduleFor...` that is exported by `ember-cli-qunit`? I'm not sure if that would even work, though, as we don't have any services/components/etc in this addon to resolve a module on.  Would it pick up on something in the dummy app, maybe?

**Comment by BrianSipple** [comment]
@elwayman02 [I just took that idea for a spin](https://github.com/BrianSipple/ember-sinon-qunit/tree/add-tests-for-each-moduleFor-variant) (which I could PR if you'd like). The dummy app is indeed a nice little test bed for each `moduleFor` variant, and more interestingly, it looks like `moduleForAcceptance` is losing the `beforeEach` context as well.

**Comment by elwayman02** [comment]
Yea, a PR for this would be great! I'd guess that if `moduleForAcceptance` doesn't work, that's an issue with `ember-cli-qunit` rather than this addon, since our approach is consistently passing the context through.

**Comment by elwayman02** [comment]
For now, you could add a failing test for `moduleForAcceptance` but use `skip` from QUnit instead of `test`, so that the test won't actually get run, but we can enable it later to see if it was fixed.

**Comment by rwjblue** [comment]
I'm happy to walk someone through this, but the issue is that the way `test` from `ember-qunit` provides `this` is not the same as the way `QUnit.test` provides `this`.  With `moduleForAcceptance` you use `QUnit.test` and with `moduleForComponent` you use `test` from `ember-qunit`.  We either need custom `test` wrapper for each environment (which sucks), or we need to only use `ember-qunit`'s `test` when running with modules from `ember-qunit`.

**Comment by elwayman02** [comment]
Honestly, the more that I think about it, the less this is an issue @BrianSipple @rwjblue. I wouldn't generally recommend using sinon for acceptance tests anyway (mirage is a better solution, imo), so just having the support for integration/unit tests is sufficient in my mind. That said, if someone really wants to use sinon for acceptance tests and puts together a proposal/PR for that functionality, I'd definitely entertain the idea of adding it here.

**Comment by mdentremont** [comment]
@elwayman02 Does mirage provide function stubbing and spying?

**Comment by BrianSipple** [comment]
@elwayman02 I agree that we can still get a lot out of sinon from the way it currently works in integration/unit tests. That said, I submitted #54 regardless, as I figured having a basic series of tests for each `module` variant can be a great way to explicitly document how we expect it to behave. 

@mdentremont [Mirage route handlers](http://www.ember-cli-mirage.com/docs/v0.2.0-beta.9/route-handlers/) can be a handy stand-in for stubbing if you're looking to simulate model data being resolved from a route during acceptance tests (or any other request, for that matter). I do see how sinon can go even further, though, when you want to do spying and what not -- so it would be interesting to see if there was _a good way_ to have a consistent pattern in the future.

**Comment by elwayman02** [comment]
@mdentremont @BrianSipple totally agreed that there are *some* use-cases for sinon in acceptance tests, but just wanted to note that the primary use-case for it overlaps heavily with mirage (server mocking). The stub/spy functionality could still be a plus, though *most* of the time we don't want to use them in acceptance tests. Often our team has found that if we don't want something to run in an acceptance test, we don't want it to run in any of them (such as analytics reporting), so we tend to create helpers that mock out those functional calls for all tests. In that case, sinon could still be used, but it would be in a test helper rather than inside each individual module, which is outside the scope of this addon (you would just import it via `ember-sinon`).

I'm still 👍 on finding a good way to have a consistent pattern for acceptance tests, but not at the cost of the current pattern that works for all of our primary use-cases. If we can make acceptance tests work without changing the public API of this addon, let's do it.

**Comment by rwjblue** [comment]
I believe that this is fixed as of https://github.com/emberjs/ember-qunit/pull/238 and https://github.com/switchfly/ember-test-helpers/pull/175.  After that commit, both acceptance and integration/unit tests use the test frameworks context.

Those changes are included in ember-qunit@0.4.21 and higher.
